File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
  let time = 0, bridge = [[0, 0]], curWeight = 0;

  while (bridge.length > 0 || truck_weights.length > 0) {
    if (bridge[0][1] === time) curWeight -= bridge.shift()[0];

    if (curWeight + truck_weights[0] <= weight) {
      curWeight += truck_weights[0];
      bridge.push([truck_weights.shift(), time + bridge_length]);
    } else {
      if (bridge[0]) time = bridge[0][1] - 1;
    }
    time++;
  }
    
  return time;
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    time = 0
    bridge = deque([0] * bridge_length)
    truck_weights = deque(truck_weights)
    
    curWeight = 0
    while len(truck_weights):
        time += 1
        curWeight -= bridge.popleft()

        if curWeight + truck_weights[0] <= weight:
            curWeight += truck_weights[0]
            bridge.append(truck_weights.popleft())
        else: bridge.append(0)

    return time + bridge_length